rfc/drafts/indexer-ideas-v1.md

```diff
@@ -0,0 +1,303 @@
+      - pictures containing same brands and trademarks as this picture
+      - more pictures in the style of this artist
+      - people with same pose as this picture
+      - same type of coffee mug as this picture
+      - more pictures of from this location
+      - more pictures of this person
+      - mashups containing this picture
+  - Or, we start with a small number of specific types of semantic similarity to start.
+  - Range of ways to support this, some relatively simple.
+
++ Eviction of images from the index?
+
++ Co-location of images that should be stored together, for faster retrieval / bulk deletion / cache promotion or eviction?
+
+
+## V1 SEARCH / DEDUPE API:
+
+
+- POST `/add`        - Add image. Image passed as base64-encoded text. Returns image ID, or, images ID'd by cryptographic hash.
+- POST `/remove`     - Mark image for removal from index.
+- POST /search     - Search by base64-encoded image or image ID. Returns list of image IDs, possibly with relevancy scores.
+- POST `/distance`   - Calculate metric-space distance between 2 images.
+- POST `/dupe_check` - Return calibrated-probability that a pair of images are duplicates.
+- GET  `/ping`       - Uptime check.
+- POST `/record_dupe`      - User labels a set of images as dupes / non-dupes. This is added to training data.
+- POST `/record_relevancy` - User labels the relevancy of a set of search results for a query. This is added to training data.
+
+
+TODO:
+
+- Fill in exact details of JSON parameters for above endpoints.
+- `/dupe_add` and `/relevancy_rate` - these are training feedback endpoints. Other types of training data and hyper-parameter settings will be also fed into the system elsewhere, but these two endpoints are included here because they may be end-user-facing?
+-   Perhaps these should be an immediate feedback tool, whose input supersedes that from the slower index re-training process?
+
+
+## SEARCH / DEDUPE ROADMAP:
+
+Version 1:
+
+  - Centralized.
+  - Off-the-shelf components.
+  - No focus on provider incentives.
+  - Unsupervised / distantly-supervised models.
+  - No custom relevancy / dedupe training.
+  - Skipping some components in V1, e.g. the "high-accuracy-model". See below.
+  - Generic wrapping APIs for each component, so they can be easily swapped out for others.
+
+Long-term:
+
+  - Distributed across machines.
+  - Distributed across users. Anyone can run an indexing node helps serve someone else's index(?)
+  - More attention to optimizing incentives among resource-providers / service-providers.
+  - Custom-trained for these particular relevancy / dedupe tasks & datasets.
+  - Accepting explicit user feedback signals(?)
+  - Mitigating attacks: resource exhaustion, bad feedback, high-trust actors gone bad, etc.
+  - Other possible ideas, see below.
+
+
+## SCALING ANN INDEXES:
+
+Distributed computing for approximate nearest neighbor (ANN) indexes requires some special consideration.
+
+Sharding:
```

Comment:
In the initial distributed approach, sharding would pseudo-randomly distribute the data evenly, among all shards in a replica set, e.g. by a simple consistent hashing function, or by something even simpler. In this version, all shards in a replica set would see all queries.

In the next iteration, the aim would be to have the shards each specialize in a local area of the search space, and queries would be routed to shards accordingly. So e.g. "random projection trees" would be a good starting basis for this routing. More advanced approaches typically combine several methods, possibly choosing methods based on the data, e.g. construct a small number of random projection trees, then do neighbor of neighbors search for each node in the tree.

Some complications arise from drift typical types of media content as the dataset grows, and rebalancing for that. 

I'll add more algorithm specifics for the RFC doc, for all of these sections.